controller/src/lib.rs

```diff
@@ -2,10 +2,605 @@
+        // part of the construction of this object.
+        let local_addr = SocketAddrV6::new(Ipv6Addr::UNSPECIFIED, PORT, 0, 0);
+        let socket = UdpSocket::bind(local_addr).await?;
+
+        // Make sure we receive addresses sent to the multicast address we use
+        // for the protocol, but not those sent by us.
+        socket.join_multicast_v6(&Ipv6Addr::from(ADDR), 0)?;
+        socket.set_multicast_loop_v6(false)?;
+
+        // Channel for communicating outgoing requests from this object to the
+        // I/O loop. Note that the _responses_ from the I/O loop back to this
+        // object are sent on a oneshot channel, which is itself placed on this
+        // channel when sending the request.
+        let (outgoing_request_tx, outgoing_request_rx) = mpsc::channel(NUM_OUTSTANDING_REQUESTS);
+
+        // Channel for communicating outgoing responses from the request-handler
+        // task to the I/O loop.
+        let (outgoing_response_tx, outgoing_response_rx) = mpsc::channel(NUM_OUTSTANDING_REQUESTS);
+
+        // Channel for communicating incoming requests from the I/O loop to
+        // the request-handler task.
+        let (incoming_request_tx, incoming_request_rx) = mpsc::channel(NUM_OUTSTANDING_REQUESTS);
+
+        // The I/O task handles the actual network I/O, reading and writing UDP
+        // packets in both directions.
+        let io_log = log.new(slog::o!("task" => "io"));
+        let io_task = tokio::spawn(async move {
+            io_loop(
+                io_log,
+                socket,
+                // For receiving requests sent from self.
+                outgoing_request_rx,
+                // For receiving responses sent from request-handler.
+                outgoing_response_rx,
+                // For sending requests from network to self.
+                incoming_request_tx,
+            )
+            .await;
+        });
+
+        // The request task runs the user-supplied request handler, receiving
+        // valid requests from the I/O task and sending valid responses back.
+        let request_log = log.new(slog::o!("task" => "request_handler"));
+        let request_task = tokio::spawn(async move {
+            request_loop(
+                request_log,
+                // For receiving requests sent from I/O loop.
+                incoming_request_rx,
+                // For sending responses to I/O loop.
+                outgoing_response_tx,
+                request_handler,
+            )
+            .await;
+        });
+
+        Ok(Self {
+            _log: log,
+            message_id: AtomicU64::new(0),
+            outgoing_request_tx,
+            io_task,
+            request_task,
+        })
+    }
+
+    // Return a header using the next available message ID.
+    fn next_header(&self) -> Header {
+        Header {
+            version: message::version::V1,
+            message_id: self.message_id.fetch_add(1, Ordering::SeqCst),
```

Comment:
Are we using an atomic here for interior mutability, or do we also need `Sync`?

If we only need interior mutability, a `Cell<u64>` might better reflect intent, e.g.
```rust
        let message_id = self.message_id.take();
        self.message_id.set(message_id + 1);
        Header {
            version: message::version::V1,
            message_id,
        }
```

(alas, [`Cell::update`](https://doc.rust-lang.org/std/cell/struct.Cell.html#method.update) remains unstable)

Reply:
My hunch is that we need `Sync`, but I'm not positive about that. The main reason is that I think this type is going to be shared between the `dpd` API server (really part of a `Switch` object it maintains), and also used by the BF SDE. The `Switch` type uses a `Mutex`, so accesses via `dpd` would be synchronized. But the BF SDE just makes FFI calls from C, so I'd expect that we to synchronize those.

Reply:
👍🏻